### PR TITLE
feat(external): add mrvi

### PR DESCRIPTION
**Pull request by martinkim0**

### CODE REVIEW

**Comment by martinkim0** [comment]
I'm going to use this branch as a staging branch and merge additional PRs into this one.

File: src/scvi/external/mrvi/_constants.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+from typing import NamedTuple
+
+
+class _MRVI_REGISTRY_KEYS_NT(NamedTuple):
+    SAMPLE_KEY: str = "sample"
```
Comment (justjhong) [inline src/scvi/external/mrvi/_constants.py:5]: let's move this into REGISTRY_KEYS

**Review by justjhong — COMMENTED** [review]
The requirement "statsmodels>=0.13.0" needs to be added (used for multiple test correction, lightweight package so shouldn't be a problem).

**Comment by martinkim0** [comment]
> The requirement "statsmodels>=0.13.0" needs to be added

Is there a particular reason for the lower bound?

**Comment by justjhong** [comment]
@martinkim0 not aware of any specific reason

**Comment by justjhong** [comment]
As mentioned offline, let's change statsmodels to scipy false_discovery_control since we already indirectly depend on scipy. https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.false_discovery_control.html

File: src/scvi/external/mrvi/_model.py [hunk 2]
```diff
@@ -0,0 +1,1495 @@
...
+                    inputs = sampled_dists
+                elif r.input == "normalized_distances":
+                    inputs = normalized_dists
+                else:
+                    raise ValueError(f"Unknown reduction input: {r.input}")
+
+                outputs = r.fn(inputs)
+
+                if r.group_by is not None:
+                    group_by = self.adata.obs[r.group_by][indices]
```
Comment (justjhong) [inline src/scvi/external/mrvi/_model.py:457]: .iloc[indices] to prevent FutureWarning spam

File: src/scvi/external/mrvi/_module.py [hunk 3]
```diff
@@ -0,0 +1,589 @@
...
+    low_dim_batch: bool = True
+    activation: Callable[[jax.typing.ArrayLike], jax.Array] = nn.gelu
+
+    @nn.compact
+    def __call__(
+        self,
+        z: jax.typing.ArrayLike,
+        batch_covariate: jax.typing.ArrayLike,
+        size_factor: jax.typing.ArrayLike,
+        continuous_covariates: jax.typing.ArrayLike | None,
```
Comment (martinkim0) [inline src/scvi/external/mrvi/_module.py:93]: This argument is currently not being used in the forward pass - is the plan to add support for this or should we remove it?

Comment (justjhong) [reply src/scvi/external/mrvi/_module.py:93]: yes let's remove, i can do this

File: CHANGELOG.md [hunk 4]
```diff
@@ -36,6 +36,7 @@ to [Semantic Versioning]. Full commit history is available in the
     {pr}`2692`.
 - Add support for custom dataloaders in {class}`scvi.model.base.VAEMixin` methods by specifying
     the `dataloader` argument {pr}`2748`.
+- Add {class}`scvi.external.MRVI` for hierarchical modeling of single-cell RNA-seq data {pr}`2756`.
```
Comment (justjhong) [inline CHANGELOG.md:39]:
```suggestion
- Add {class}`scvi.external.MRVI` for modeling sample-level heterogeneity in single-cell RNA-seq data {pr}`2756`.
```

File: src/scvi/utils/_docstrings.py [hunk 5]
```diff
@@ -118,6 +118,12 @@
     integer categories and saved to `adata.obs['_scvi_batch']`. If `None`, assigns the same batch
     to all the data."""
 
+param_sample_key = """\
+sample_key
+    key in `adata.obs` for sample information. Categories will automatically be converted into
+    integer categories and saved to `adata.obs['_scvi_sample']`. If `None`, assigns the same batch
```
Comment (justjhong) [inline src/scvi/utils/_docstrings.py:124]:
```suggestion
    integer categories and saved to `adata.obs['_scvi_sample']`. If `None`, assigns the same sample
```

**Review by justjhong — APPROVED** [review]
Looks great!! Thanks for all the work 👍  As discussed, let's hold off until we have an idea of the timeline

File: src/scvi/external/mrvi/_module.py [hunk 6]
```diff
@@ -0,0 +1,579 @@
...
+    n_input
+        Number of input features.
+    n_sample
+        Number of samples.
+    n_batch
+        Number of batches.
+    n_labels
+        Number of labels.
+    n_latent
+        Number of latent variables.
```
Comment (justjhong) [inline src/scvi/external/mrvi/_module.py:300]:
```suggestion
        Number of latent variables for ``z``.
```

File: src/scvi/external/mrvi/_module.py [hunk 7]
```diff
@@ -0,0 +1,579 @@
...
+
+        qz_kwargs = DEFAULT_QZ_ATTENTION_KWARGS.copy()
+        if self.qz_kwargs is not None:
+            qz_kwargs.update(self.qz_kwargs)
+
+        qu_kwargs = DEFAULT_QU_KWARGS.copy()
+        if self.qu_kwargs is not None:
+            qu_kwargs.update(self.qu_kwargs)
+
+        is_isomorphic_uz = self.n_latent == self.n_latent_u
```
Comment (justjhong) [inline src/scvi/external/mrvi/_module.py:370]: I'd like to raise a warning if n_latent is set lower than n_latent_u. what's the best way to do this @martinkim0 ?

Comment (martinkim0) [reply src/scvi/external/mrvi/_module.py:370]: I can take a look at this!

**Pull request merged**